nixosTests.kexec: do not use bootloader

RaitoBezarius opened the pull request:
###### Description of changes

There's no reason to use a bootloader when testing kexec, this is a feature that reboots *directly* in the kernel, if anything, we should just direct boot the kernel and reboots in the kernel.

~A bootloader test really makes sense to test "default" systemctl kexec behavior which is already broken because systemctl kexec will read the ESP to determine what to kexec by default.~

It still fails because there is no proper reconnection in the second part of the test, though. ~Please maintainers take a look.~ I'm doing that now. :P

This can unfortunately be related to a shadow mounting of the Nix store, in which case, some fixes are around in my PRs.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
nixosTests.kexec: do not use bootloader

There's no reason to use a bootloader when testing kexec, this is a feature
that reboots *directly* in the kernel, if anything, we should just direct boot the
kernel and reboots in the kernel.

A bootloader test really makes sense to test "default" systemctl kexec behavior which is already broken
because systemctl kexec will read the ESP to determine what to kexec by default.

flokli commented:
Was this just copypasted from another test, or why was it in there initially?

RaitoBezarius commented:
introduced in 6ceedff33160

ElvishJerricco commented:
> A bootloader test really makes sense to test "default" systemctl kexec behavior which is already broken because systemctl kexec will read the ESP to determine what to kexec by default.

I thought we determined (while discussing #232660) that this is actually correct, because it means that `systemctl kexec` will boot the same exact things from the ESP that you would boot if you did a real reboot, respecting all the EFI variables that systemd-boot respects like `bootctl set-oneshot`? The only issue is that it doesn't work with near-UKIs like lanzaboote.

Not relevant for this test as-written though, since it manually loads a kernel/initrd/cmdline before calling `systemctl kexec`. So it's still weird that this test is failing with this change.

RaitoBezarius commented:
> > A bootloader test really makes sense to test "default" systemctl kexec behavior which is already broken because systemctl kexec will read the ESP to determine what to kexec by default.
> 
> I thought we determined (while discussing #232660) that this is actually correct, because it means that `systemctl kexec` will boot the same exact things from the ESP that you would boot if you did a real reboot, respecting all the EFI variables that systemd-boot respects like `bootctl set-oneshot`? The only issue is that it doesn't work with near-UKIs like lanzaboote.
> 
> Not relevant for this test as-written though, since it manually loads a kernel/initrd/cmdline before calling `systemctl kexec`. So it's still weird that this test is failing with this change.

Ahem, you're right.

I'm debugging it and this basically boils down to "I don't even know how this test worked" before :P.
When kexecing, the connection to the VM gets cut off, you need to reconnect, on the first time, this is done.

On the second time, this is not. Some garbage appears in the output buffer "Spawning a backdoor shell" which is a not a valid base64 string which is expected by those `succeed/execute` functions. Failing the whole pipeline.

The solution is to properly reconnect and drain the buffer at kexec time to avoid getting those messages for nothing.

RaitoBezarius committed:
nixosTests.kexec: reconnect properly after 2nd kexec

By some miracle, before, it was possible to reconnect to the `node1` without
doing any relevant dance.

But now we are direct booting (¿), it seems like we need to do the right things.

This introduces a `check_output` flag for `execute` because we do not want to steal the
messages from the backdoor service as we might execute the kexec too fast compared
to when we will reconnect.

Therefore, we will let the message in the pipe if needed.

RaitoBezarius commented:
@ofborg test kexec

RaitoBezarius commented:
@tfc can you look over for the reconnection solution? I'm surprised we need to do this and at the same time… I guess that's normal?

RaitoBezarius committed:
nixosTests.kexec: do not use module argument `config`

It is deprecated now.

RaitoBezarius commented:
@tfc deprecation is addressed, can you formally bless the PR? :P

Artturin commented:
Please use your own fork for your pr branches
